### PR TITLE
Add pluginmanager fixture

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
Add `napari_plugin_tester` fixture.  provides a few builtin test methods, and we can expand this as desired (cc @ziyangczi).

As an example, a plugin developer would use this as follows:

```py
# assume the plugin has registered their module `_test.py`
# under the plugin name `test-plugin`
# and it includes a single hookimpl for napari_get_reader

import _test


def test_plugin_registration(napari_plugin_tester):
    napari_plugin_tester.assert_plugin_name_registered("test-plugin")
    napari_plugin_tester.assert_module_registered(_test)
    napari_plugin_tester.assert_implementations_registered(
        "test-plugin", "napari_get_reader"
    )

```

elaborations might include "autodetecting" the module in which the test is run, in order to either parse the source and provide intelligent feedback. Or perhaps making sure that other plugins installed in the environment don't influence the test, etc...

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by sofroniewn — APPROVED** [review]
LGTM! @ziyangczi do you want to review before I merge? Might be nice to make this available for 0.4.5

File: napari/_tests/test_pytest_plugin.py [hunk 1]
```diff
@@ -30,3 +30,32 @@ def test_make_viewer(make_napari_viewer):
...
+            napari_plugin_tester.register(Plugin)
+            napari_plugin_tester.assert_plugin_name_registered("Plugin")
+            napari_plugin_tester.assert_module_registered(Plugin)
+            napari_plugin_tester.assert_implementations_registered(
+                Plugin, "napari_get_reader"
+            )
+    """
+    )
+    # run all tests with pytest
+    result = testdir.runpytest()
```
Comment (jni) [inline napari/_tests/test_pytest_plugin.py:58]: Jeez sometimes I really hate pytest magic... Is testdir a magical fixture that you get automatically with pytest?

Comment (tlambert03) [reply napari/_tests/test_pytest_plugin.py:58]: I know... it comes from pytest itself, and is provided to test plugins: https://docs.pytest.org/en/stable/writing_plugins.html#testing-plugins

Comment (jni) [reply napari/_tests/test_pytest_plugin.py:58]: I just saw that pytester is only available in 6.2 (?) — do we want to add that minimum version to our setup.cfg?

Comment (tlambert03) [reply napari/_tests/test_pytest_plugin.py:58]: sure I can add that

File: napari/utils/_testsupport.py [hunk 2]
```diff
@@ -69,3 +70,59 @@ def actual_factory(*model_args, viewer_class=Viewer, **model_kwargs):
...
+
+    See _TestPluginManager above for tests implementations:
+    Examples
+    --------
+    >>> def test_pm(napari_plugin_tester):
+    >>> napari_plugin_tester.assert_plugin_name_registered("test-plugin")
+    >>> napari_plugin_tester.assert_module_registered(_test)
+    >>> napari_plugin_tester.assert_implementations_registered(
+    >>>     "test-plugin", "napari_get_reader"
+    >>> )
```
Comment (jni) [inline napari/utils/_testsupport.py:121]:
```suggestion
    >>> def test_pm(napari_plugin_tester):
    ...     napari_plugin_tester.assert_plugin_name_registered("test-plugin")
    ...     napari_plugin_tester.assert_module_registered(_test)
    ...     napari_plugin_tester.assert_implementations_registered(
    ...         "test-plugin", "napari_get_reader"
    ...     )
```

File: napari/utils/_testsupport.py [hunk 3]
```diff
@@ -69,3 +70,59 @@ def actual_factory(*model_args, viewer_class=Viewer, **model_kwargs):
...
+                        f"{hook!r} was not registered for plugin {plugin!r}"
+                    )
+
+
+@pytest.fixture
+def napari_plugin_tester():
+    """A fixture that can be used to test plugin registration.
+
+    See _TestPluginManager above for tests implementations:
+    Examples
```
Comment (jni) [inline napari/utils/_testsupport.py:114]:
```suggestion

    Examples
```

**Comment by jni** [comment]
Summarising a quick discussion with @tlambert03 on Zulip:

- pytester [docs](https://docs.pytest.org/en/stable/reference.html#pytester) say it's only available on pytest 6.2, but @tlambert03 can get this to work on 6.0. Turns out the pytester *plugin* has been around since 5.0 but the builtin magical fixture only since 6.2. We can use the plugin for now.
- the testdir [docs](https://docs.pytest.org/en/stable/reference.html#testdir) say "New code should avoid using testdir in favor of pytester.", but we can punt on this, since it won't change the API of our own fixture, which is nice.
- I also think the fixture methods should get docstrings but we can also punt on that.

File: napari/_tests/test_pytest_plugin.py (hunk 1 above)
Comment (tlambert03) [reply napari/_tests/test_pytest_plugin.py:58]: as discussed on zulip, "pytester" the _fixture_ is new as of 6.2 ... but pytester the _plugin_ (with the `testdir` fixture) has been around for a long time.

**Pull request merged**